lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,418 @@
+        continue;
+      auto operandInst = dyn_cast<Instruction>(use.get());
+      if (!operandInst)
+        continue;
+      if (m_canAssumeUniformDivergentUsesMap.count(operandInst) == 1)
+        continue;
+      if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block))
+        continue;
+      // The operand can be assumed uniform becasue all users are assumed uniform
+      if (isAllUsersAssumedUniform(operandInst)) {
+        m_canAssumeUniformDivergentUsesMap[operandInst] = {};
+        m_canAssumeUniforms.push_back(operandInst);
+      }
+    }
+
+    bool canCollectOperands = true;
+    // NOTE: Stop propagation for a loadInst whose address space is fat pointer because it has uncertain size
+    auto loadInst = dyn_cast<LoadInst>(&inst);
+
+    // For all operands, see if all operands can be assumed uniform and are in the same block. If true, we can collect
+    // the operands to be used for the propagation heuristic in findInsertLocation.
+    SmallVector<Instruction *, 2> operandInsts;
+    for (Use &use : inst.operands()) {
+      if (auto operandInst = dyn_cast<Instruction>(use.get())) {
+        if (!m_canAssumeUniformDivergentUsesMap.count(operandInst))
+          continue;
+        if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block) ||
+            (loadInst && loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER)) {
+          canCollectOperands = false;
+          break;
+        }
+        operandInsts.push_back(operandInst);
+      }
+    }
+    if (canCollectOperands)
+      m_canAssumeUniformDivergentUsesMap[&inst] = std::move(operandInsts);
+  }
+}
+
+// =====================================================================================================================
+// Find the best insert locations according to the heuristic.
+// The heuristic is: if an instruction that can be assumed to be uniform has multiple divergent operands, then you take
+// the definition of the divergent operand that is earliest in basic block order (call it "the earliest divergent
+// operand") and propagate up to that instruction; if it turns out that that instruction can be assumed to be uniform,
+// then we can just insert the readfirstlane there (or propagate).
+//
+// @param block : The process basic block
+// @param readFirstLaneCount : The initial amdgcn_readfirstlane vector
+void PatchReadFirstLane::findInsertLocations(BasicBlock *block,
+                                             const SmallVectorImpl<Instruction *> &initialReadFirstLanes) {
+  m_insertLocations.clear();
+
+  // The principle is to lift readfirstlane to a best insert location if it exists but not introduce extra
+  // readfirstlane.
+  DenseSet<Instruction *> visitedCandidates;
+  for (auto &initialReadFirstLane : initialReadFirstLanes) {
+    // Find an insert location for each initial readfirstlane
+    std::deque<Instruction *> candidateInstQueuqe;
```

Comment:
Name seems a bit strange: candidateInstDeque?

Reply:
Fixed the typo.